File: diagnostics/j2534-host-node/capability-token.js
/**
 * ECDSA P-256 (ES256) diagnostics capability tokens (format `v2.<payload>.<sig>`).
 *
 * The Cloudflare Worker signs tokens with a private key that never leaves the
 * server. Hosts hold only the PUBLIC key, so a compromised or reverse-engineered
 * client cannot mint tokens (fixes the previous shared-HMAC-secret weakness).
 *
 * Fail-closed: verification requires a configured public key. There is no
 * built-in key that could validate real tokens.
 * Compatible with diagnostics/j2534-service/src/J2534.Host/CapabilityToken.cs.
 */
const { createPublicKey, createPrivateKey, sign, verify, randomBytes } = require('node:crypto');
const fs = require('node:fs');
const path = require('node:path');

/** jti -> expiry (ms). Single-use enforcement with expiry-aware pruning. */
const consumed = new Map();

function pemFromSpkiBase64(base64) {
  const body = base64.replace(/\s+/g, '').match(/.{1,64}/g).join('\n');
  return `-----BEGIN PUBLIC KEY-----\n${body}\n-----END PUBLIC KEY-----\n`;
}

function defaultPublicKeyPaths() {
  return [
    // Ephemeral dev key written by .cursor/install.sh (gitignored).
    path.join(__dirname, '..', 'keys', 'local-capability-public-key.pem'),
    // Production key provisioned by CI alongside the packaged build.
    path.join(__dirname, '..', 'keys', 'capability-public-key.pem'),
  ];
}

function loadPublicKey() {
  const inline = String(process.env.DIAGNOSTICS_SIGNING_PUBLIC_KEY || '').trim();
  if (inline) {
    return createPublicKey(inline.includes('BEGIN') ? inline : pemFromSpkiBase64(inline));
  }
  const explicitFile = String(process.env.DIAGNOSTICS_SIGNING_PUBLIC_KEY_FILE || '').trim();
  const candidates = explicitFile ? [explicitFile, ...defaultPublicKeyPaths()] : defaultPublicKeyPaths();
  for (const candidate of candidates) {
    try {
      if (candidate && fs.existsSync(candidate)) return createPublicKey(fs.readFileSync(candidate, 'utf8'));
    } catch { /* try next candidate */ }
  }
  return null;
}

function pruneConsumed() {
  const now = Date.now();
  for (const [jti, exp] of consumed) {
    if (exp <= now) consumed.delete(jti);
  }
}

/**
 * Verify a capability token. `expected` may constrain procedure, vin, and mode.
 * Pass `consume: false` for a pre-flight signature/expiry check that does not
 * burn the single-use nonce (the executing host performs the consuming check).
 */
function verifyCapabilityToken(token, expected = {}) {
  const raw = String(token || '').trim();
  const parts = raw.split('.');
  if (parts.length !== 3 || parts[0] !== 'v2') {
    throw new Error('Invalid diagnostics capability token');
  }
  const publicKey = loadPublicKey();
  if (!publicKey) {
    throw new Error('Diagnostics capability public key is not configured; refusing to authorize.');
  }
  const signingInput = `${parts[0]}.${parts[1]}`;
  const signature = Buffer.from(parts[2], 'base64url');
  const valid = verify('sha256', Buffer.from(signingInput), { key: publicKey, dsaEncoding: 'ieee-p1363' }, signature);
  if (!valid) {
    throw new Error('Invalid diagnostics capability token signature');
  }
  const payload = JSON.parse(Buffer.from(parts[1], 'base64url').toString('utf8'));
  if (payload.v !== 2 || !payload.procedure || !payload.vin || !payload.shopId || !payload.jti || !payload.exp || !payload.mode) {
    throw new Error('Capability token payload is incomplete');
  }
  if (Date.now() > Number(payload.exp)) {
    throw new Error('Capability token expired');
  }
  if (expected.procedure && payload.procedure !== expected.procedure) {
    throw new Error(`Capability token is not valid for ${expected.procedure}`);
  }
  if (expected.vin && payload.vin !== String(expected.vin).trim().toUpperCase()) {
    throw new Error('Capability token VIN mismatch');
  }
  if (expected.mode && payload.mode !== expected.mode) {
    throw new Error('Capability token mode mismatch');
  }
  if (expected.consume !== false) {
    pruneConsumed();
    if (consumed.has(payload.jti)) {
      throw new Error('Capability token already used');
    }
    consumed.set(payload.jti, Number(payload.exp));
  }
  return payload;
}

/**
 * Sign a capability token with an ECDSA P-256 private key (PEM/DER). Intended
 * for development seeding and tests only — production tokens are minted by the
 * Worker, which holds the private key.
 */
function mintCapabilityToken(privateKey, input) {
  const key = typeof privateKey === 'string' && !privateKey.includes('BEGIN')
    ? createPrivateKey({ key: Buffer.from(privateKey, 'base64'), format: 'der', type: 'pkcs8' })
    : createPrivateKey(privateKey);
  const now = Date.now();
  const payload = {
    v: 2,
    procedure: input.procedure,
    scope: input.scope || '',
    vin: String(input.vin || '').trim().toUpperCase(),
    shopId: input.shopId || 'local',
    mode: input.mode || 'simulate',
    actor: input.actor || null,
    iat: now,
    exp: now + (input.ttlMs || 5 * 60 * 1000),
    jti: randomBytes(12).toString('hex'),
  };
  const payloadB64 = Buffer.from(JSON.stringify(payload)).toString('base64url');
  const signingInput = `v2.${payloadB64}`;
  const signature = sign('sha256', Buffer.from(signingInput), { key, dsaEncoding: 'ieee-p1363' }).toString('base64url');
  return { token: `${signingInput}.${signature}`, payload };
}

module.exports = { verifyCapabilityToken, mintCapabilityToken, loadPublicKey };

File: diagnostics/j2534-service/src/J2534.Host/CapabilityToken.cs
using System.Collections.Concurrent;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace MechPro.J2534.Host;

/// <summary>
/// ECDSA P-256 (ES256) capability tokens (format v2.&lt;payload&gt;.&lt;sig&gt;).
///
/// The Cloudflare Worker signs with a private key that never leaves the server;
/// this host holds only the PUBLIC key and can verify but not mint tokens.
/// Fail-closed: without a configured public key, every authorized procedure is
/// refused. Compatible with the Node verifier and Worker minter.
/// </summary>
public static class CapabilityToken
{
    static readonly ConcurrentDictionary<string, long> Consumed = new();
    static readonly Lazy<ECDsa?> PublicKey = new(LoadPublicKey);

    public sealed record Payload(string Procedure, string Vin, string ShopId, string Mode, string Scope);

    /// <summary>Verify a token scoped to <paramref name="procedure"/> and (optionally) mode/vin. Single-use unless consume=false.</summary>
    public static Payload Verify(string? token, string procedure, string? expectedMode = null, string? expectedVin = null, bool consume = true)
    {
        var raw = (token ?? string.Empty).Trim();
        var parts = raw.Split('.');
        if (parts.Length != 3 || parts[0] != "v2")
        {
            throw new UnauthorizedAccessException("Invalid diagnostics capability token");
        }

        var key = PublicKey.Value
            ?? throw new UnauthorizedAccessException("Diagnostics capability public key is not configured; refusing to authorize.");

        var signingInput = Encoding.UTF8.GetBytes($"{parts[0]}.{parts[1]}");
        var signature = Base64UrlDecode(parts[2]);
        if (!key.VerifyData(signingInput, signature, HashAlgorithmName.SHA256))
        {
            throw new UnauthorizedAccessException("Invalid diagnostics capability token signature");
        }

        using var doc = JsonDocument.Parse(Encoding.UTF8.GetString(Base64UrlDecode(parts[1])));
        var root = doc.RootElement;
        if (root.GetProperty("v").GetInt32() != 2)
        {
            throw new UnauthorizedAccessException("Unsupported capability token version");
        }

        var proc = root.GetProperty("procedure").GetString();
        var vin = root.GetProperty("vin").GetString();
        var shopId = root.GetProperty("shopId").GetString();
        var mode = root.GetProperty("mode").GetString();
        var scope = root.TryGetProperty("scope", out var s) ? s.GetString() ?? "" : "";
        var jti = root.GetProperty("jti").GetString();
        var exp = root.GetProperty("exp").GetInt64();

        if (string.IsNullOrWhiteSpace(proc) || string.IsNullOrWhiteSpace(vin)
            || string.IsNullOrWhiteSpace(shopId) || string.IsNullOrWhiteSpace(mode) || string.IsNullOrWhiteSpace(jti))
        {
            throw new UnauthorizedAccessException("Capability token payload is incomplete");
        }
        if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() > exp)
        {
            throw new UnauthorizedAccessException("Capability token expired");
        }
        if (!string.Equals(proc, procedure, StringComparison.Ordinal))
        {
            throw new UnauthorizedAccessException($"Capability token is not valid for {procedure}");
        }
        if (expectedMode is not null && !string.Equals(mode, expectedMode, StringComparison.Ordinal))
        {
            throw new UnauthorizedAccessException("Capability token mode mismatch");
        }
        if (expectedVin is not null && !string.Equals(vin, expectedVin.Trim().ToUpperInvariant(), StringComparison.Ordinal))
        {
            throw new UnauthorizedAccessException("Capability token VIN mismatch");
        }

        if (consume)
        {
            PruneConsumed();
            if (!Consumed.TryAdd(jti!, exp))
            {
                throw new UnauthorizedAccessException("Capability token already used");
            }
        }

        return new Payload(proc!, vin!, shopId!, mode!, scope);
    }

    static void PruneConsumed()
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        foreach (var kv in Consumed)
        {
            if (kv.Value <= now) Consumed.TryRemove(kv.Key, out _);
        }
    }

    static ECDsa? LoadPublicKey()
    {
        var inline = Environment.GetEnvironmentVariable("DIAGNOSTICS_SIGNING_PUBLIC_KEY");
        if (!string.IsNullOrWhiteSpace(inline))
        {
            return ImportPublicKey(inline);
        }

        var candidates = new List<string>();
        var explicitFile = Environment.GetEnvironmentVariable("DIAGNOSTICS_SIGNING_PUBLIC_KEY_FILE");
        if (!string.IsNullOrWhiteSpace(explicitFile)) candidates.Add(explicitFile);
        var baseDir = AppContext.BaseDirectory;
        candidates.Add(Path.Combine(baseDir, "diagnostics-keys", "capability-public-key.pem"));
        candidates.Add(Path.Combine(baseDir, "capability-public-key.pem"));

        foreach (var candidate in candidates)
        {
            try
            {
                if (File.Exists(candidate)) return ImportPublicKey(File.ReadAllText(candidate));
            }
            catch { /* try next candidate */ }
        }
        return null;
    }

    static ECDsa ImportPublicKey(string material)
    {
        var ecdsa = ECDsa.Create();
        if (material.Contains("BEGIN"))
        {
            ecdsa.ImportFromPem(material);
        }
        else
        {
            ecdsa.ImportSubjectPublicKeyInfo(Convert.FromBase64String(material.Trim()), out _);
        }
        return ecdsa;
    }

    static byte[] Base64UrlDecode(string input)
    {
        var padded = input.Replace('-', '+').Replace('_', '/');
        switch (padded.Length % 4)
        {
            case 2: padded += "=="; break;
            case 3: padded += "="; break;
        }
        return Convert.FromBase64String(padded);
    }
}
